lib/y-richtext.coffee

```diff
@@ -83,81 +99,104 @@ class RichText extends BaseClass
-          else if event.type == "delete"
-            #FIXME: depending on how it is implemented, selections can be
-            # overridden by negating their values (set them to null) or they
-            # can just be deleted
-            console.log "Ohow… what am I supposed to do there?"
+  # TODO: should be private!
+  bindEventsToEditor : (editor) ->
+    # update the editor when something on the $cursors happens
+    ###
+    @_get("cursors").observe (events)=>
+      if this.lock_editor_propagation
+        # break, if lock is on
+        return
+      this.lock_editor_propagation = true
+      for event in events
+        id = event.name
+        index = event.object.val(event.name).getPosition()
+        text = event.name
+        color = "grey" # FIXME
+
+        @editor.setCursor id, index, text, color
+      this.lock_editor_propagation = false
+    ###
+
+    # update the editor when something on the $characters happens
+    @_get("characters").observe (events)=>
+      if this.lock_editor_propagation
+        # break, if lock is on
+        return
+      this.lock_editor_propagation = true
+      for event in events
+        delta =
+          ops: [{retain: event.position}]
+
+        if event.type == "insert"
+          delta.ops.push {insert: event.value}
+
+        else if event.type == "delete"
+          delta.ops.push {delete: 1}
+
+        @editor.updateContents delta
+      this.lock_editor_propagation = false
+
+    # update the editor when something on the $selections happens
+    @_get("selections").observe (event)=>
+      if this.lock_editor_propagation
+        # break, if lock is on
+        return
+      this.lock_editor_propagation = true
+
+      attrs = {}
+      if event.type is "select"
+        for attr,val of event.attrs
+          attrs[attr] = val
+      else # is "unselect"!
+        for attr in event.attrs
+          attrs[attr] = null
+      retain = event.from.getPosition()
+      selection_length = event.to.getPosition()-event.from.getPosition()
+      @editor.updateContents
+        ops: [
+          {retain: retain},
+          {retain: selection_length, attributes: attrs}
+        ]
+
+      this.lock_editor_propagation = false
 
   # Apply a delta and return the new position
   # @param delta [Object] a *single* delta (see ot-types for more info)
   # @param position [Integer] start position for the delta, default: 0
   #
   # @return [Integer] the position of the cursor after parsing the delta
-  deltaHelper = (delta, position =0) ->
+  deltaHelper : (delta, position = 0) ->
     if delta?
-      arentNull = (el) -> el != null
       selections = (@_get "selections")
-      if _.all delta.attributes, arentNull
-        operation = selections.select
-      else
-        operation = selections.unselect
+      delta_unselections = []
+      delta_selections = {}
+      for n,v of delta.attributes
+        if v?
+          delta_selections[n] = v
+        else
+          delta_unselections.push n
```

Comment:
I don't know to what you are refering to. Do you mean *n*, and *v*? I figure it is quite alright in this case, because putting `attribute_name` and `attribute_value` here does not add much information to the code. If you have a really good alternative, tell me!